sycl/doc/SYCLPluginInterface.md

```diff
@@ -0,0 +1,178 @@
+  library and call it with the string *“clIcdGetPlatformIDsKHR”* to find
+  the ```clIcdGetPlatformIDsKHR``` function pointer
+- then, using ```clIcdGetPlatformIDsKHR```, find the set of
+  ```cl_platform_id``` objects that the library provides, and store them
+  internally to enumerate available devices when required
+
+### Device enumeration by plugins
+
+TBD
+
+## PI API Specification
+
+PI allows for two levels of API implementation - basic and extended. The basic
+level comprises core APIs needed to perform SYCL offload. The extended level is
+needed only for OpenCL interoperability. 
+
+### The basic PI APIs
+
+This level requires availability of functions representing core functionality,
+such as device memory management, kernel creation and parameter setting,
+enqueuing kernel for execution, etc. There are functions which fully match
+semantics of the corresponding OpenCL X.X functions, as well as functions which
+don't have direct mapping to OpenCL. See the
+[pi.h](../include/CL/sycl/detail/pi.h) header for the full list and
+descriptions of PI APIs. [TBD: link to pi.h doxygen here]
+
+*__Examples of PI OpenCL-based APIs__*
+
+    piCreateKernel
+    piReleaseKernel
+    piSetKernelArg
+
+*__Example of PI non OpenCL-based APIs__*
+
+Selects the most appropriate device image based on runtime
+information and the image characteristics
+```
+cl_int piSelectDeviceImage(pi_context ctx, pi_device_image **images,
+                             cl_uint num_images,
+                             pi_device_image **selected_image);
+```
+
+PI also defines few types and string tags to describe a device binary image.
+Those are used to communicate to plugins information about the images where it
+is needed, currently only in the ```piSelectDeviceImage``` function. The main
+type is ```pi_device_image```, whose detailed description can also be found in
+the header.  The layout of this type strictly matches the layout of the corresponding
+device image descriptor type defined in the ```clang-offload-wrapper``` tool
+which wraps device binaries into a host object for further linkage. The wrapped
+images reside inside this descriptor in a data section.
+
+### The Extended PI APIs
+
+These are APIs needed to implement SYCL-to-OpenCL interoperability. Only
+the OpenCL plugin will provide these APIs, generic plugins won't, because
+SYCL runtime will cast PI object pointers to/from OpenCL object pointers when
+using result of/invoking the extended APIs in interoperability scenario. These
+APIs match semantics of the corresponding OpenCL APIs exactly.
+
+The full list of such APIs can be found in the
+[pi.h](../include/CL/sycl/detail/pi.h) header. One example is:
+
+    piOpenCLCreateProgramWithSource
+ 
+### Extension mechanism
+
+This section describes a mechanism for SYCL or other runtimes to detect
+availability of and obtain interfaces beyond those defined by the PI dispatch.
+The ```piGetExtensionFunctionAddressForPlatform``` function can be used
```

Comment:
Why can't we just define more PI interfaces and make them optional to implement?  Unlike OpenCL, the PI API is not a standard.

Reply:
That makes sense, but also means some kind of agreement for everyone using the PI interface will need to be in place as to what is an extension to PI and what is a core PI functionality - hence why I suggested in a different place to at least have a version number.
If we take what is in this repo SYCL branch as the golden PI interface, pull requests can be used to add new functionality.

Reply:
OK, I'll remove this one leaving TBD for the extension mechanism in this place. Originally PI was design to be binary compatible with OpenCL so that OpenCL dispatch tables could be used in place of PI dispatch table, now the approach has evolved.